File: VB/MasterDetailPerformance/CustomTableView.vb
Imports System.Collections
Imports System.Linq
Imports DevExpress.Xpf.Grid
Imports DevExpress.Xpf.Grid.Printing

Namespace MasterDetailPerformance
    Public Class CustomTableView
        Inherits TableView

        Protected Overrides Function CreatePrintingDataTreeBuilder(ByVal totalHeaderWidth As Double, ByVal itemsGenerationStrategy As ItemsGenerationStrategyBase, ByVal masterDetailPrintInfo As MasterDetailPrintInfo, ByVal bandsLayout As BandsLayoutBase) As PrintingDataTreeBuilderBase
            Return MyBase.CreatePrintingDataTreeBuilder(totalHeaderWidth, New ItemsGenerationTop10RowsStrategy(ViewInfo.GridView), masterDetailPrintInfo, bandsLayout)
        End Function
    End Class
    Public Class ItemsGenerationTop10RowsStrategy
        Inherits ItemsGenerationServerStrategy

        Public Sub New(ByVal view As DataViewBase)
            MyBase.New(view)
        End Sub
        Public Overrides ReadOnly Property RequireFullExpand() As Boolean
            Get
                Return True
            End Get
        End Property

        Private _SelectedRecordsNumber As Integer = 10

        Public Property SelectedRecordsNumber() As Integer
            Get
                Return _SelectedRecordsNumber
            End Get
            Set(ByVal value As Integer)
                _SelectedRecordsNumber = value
            End Set
        End Property
        Protected Overrides Function FetchAllFilteredAndSortedRows() As IList
            Return DirectCast(View.DataControl.ItemsSource, IEnumerable).Cast(Of Object)().Take(SelectedRecordsNumber).ToList()
        End Function

        Protected Overrides Function GetAllFilteredAndSortedRows() As IList
            Return FetchAllFilteredAndSortedRows()
        End Function
    End Class
End Namespace

File: CS/MasterDetailPerformance/CustomTableView.cs
using System.Collections;
using System.Linq;
using DevExpress.Xpf.Grid;
using DevExpress.Xpf.Grid.Printing;

namespace MasterDetailPerformance {
    public class CustomTableView : TableView {
        protected override PrintingDataTreeBuilderBase CreatePrintingDataTreeBuilder(double totalHeaderWidth, ItemsGenerationStrategyBase itemsGenerationStrategy, MasterDetailPrintInfo masterDetailPrintInfo, BandsLayoutBase bandsLayout) {
            return base.CreatePrintingDataTreeBuilder(totalHeaderWidth, new ItemsGenerationTop10RowsStrategy(ViewInfo.GridView), masterDetailPrintInfo, bandsLayout);
        }
    }
    public class ItemsGenerationTop10RowsStrategy : ItemsGenerationServerStrategy {
        public ItemsGenerationTop10RowsStrategy(DataViewBase view) : base(view) { }
        public override bool RequireFullExpand { get { return true; } }

        private int _SelectedRecordsNumber = 10;

        public int SelectedRecordsNumber {
            get {
                return _SelectedRecordsNumber;
            }
            set {
                _SelectedRecordsNumber = value;
            }
        }
        protected override IList FetchAllFilteredAndSortedRows() {
            return ((IEnumerable)View.DataControl.ItemsSource)
                .Cast<object>()
                .Take(SelectedRecordsNumber)
                .ToList();
        }

        protected override IList GetAllFilteredAndSortedRows() {
            return FetchAllFilteredAndSortedRows();
        }
    }
}
